overlord/snapstate/handlers.go

```diff
@@ -607,8 +608,17 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	}
 	// only set userID if unset in snapst and if we actually have
 	// an associated user
-	if snapst.UserID == 0 && snapsup.UserID > 0 {
-		snapst.UserID = snapsup.UserID
+	if snapsup.UserID > 0 {
+		var user *auth.UserState
+		if snapst.UserID != 0 {
+			user, err = auth.User(st, snapst.UserID)
+			if err != nil && err != auth.ErrInvalidUser {
+				return err
+			}
+		}
+		if user == nil {
```

Comment:
Maybe a small comment here like: `if the original user installing the snap is no longer available transfer to to user who triggered this change` (or similar :)